drivers/flash/nrf_qspi_nor.c

```diff
@@ -0,0 +1,673 @@
+				return -EIO;
+			}
+
+			u8_t tx = BIT(CONFIG_NORDIC_QSPI_NOR_QE_BIT);
+
+			const struct qspi_buf tx_buff = { .buf = &tx, .len = sizeof(tx), };
+
+			cmd.op_code = SPI_NOR_CMD_WRSR;
+			cmd.tx_buf = &tx_buff;
+			cmd.rx_buf = NULL;
+
+			if (qspi_send_cmd(dev, &cmd) != 0) {
+				return -EIO;
+			}
+		}
+	}
+
+	return qspi_get_zephyr_ret_code(res);
+}
+
+/**
+ * @brief Retrieve the Flash JEDEC ID and compare it with the one expected
+ *
+ * @param dev The device structure
+ * @param flash_id The flash info structure which contains the
+ *		  expected JEDEC ID
+ * @return 0 on success, negative errno code otherwise
+ */
+static inline int qspi_nor_read_id(struct device *dev,
+				   const struct qspi_nor_config *const flash_id)
+{
+	u8_t rx_b[QSPI_NOR_MAX_ID_LEN];
+	const struct qspi_buf q_rx_buf = {
+		.buf = rx_b,
+		.len = QSPI_NOR_MAX_ID_LEN
+	};
+	const struct qspi_cmd cmd = {
+		.op_code = SPI_NOR_CMD_RDID,
+		.rx_buf = &q_rx_buf,
+		.tx_buf = NULL
+	};
+
+	if (qspi_send_cmd(dev, &cmd) != 0) {
+		return -EIO;
+	}
+
+	if (memcmp(flash_id->id, rx_b, QSPI_NOR_MAX_ID_LEN) != 0) {
+		LOG_ERR("flash id error. Extected: [%d %d %d], got: [%d %d %d]",
+			flash_id->id[0], flash_id->id[1], flash_id->id[2],
+			rx_b[0], rx_b[1], rx_b[2]);
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int qspi_nor_read(struct device *dev, off_t addr, void *dest,
+			 size_t size)
+{
+	if (!dest) {
+		return -EINVAL;
+	}
+
+	/* read size must be multiple of 4 bytes */
+	if (size % sizeof(uint32_t) ||
+	    !(size > 0)) {
+		return -EINVAL;
+	}
+
+	const struct qspi_nor_config *params = dev->config->config_info;
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((addr + size) > params->size)) {
+		return -EINVAL;
+	}
+
+	qspi_lock(dev);
+
+	int ret = nrfx_qspi_read(dest, size, addr);
+
+	qspi_wait_for_completion(dev);
+	return qspi_get_zephyr_ret_code(ret);
+}
+
+static int qspi_nor_write(struct device *dev, off_t addr, const void *src,
+			  size_t size)
+{
+	if (!src) {
+		return -EINVAL;
+	}
+
+	/* write size must be multiple of 4 bytes */
+	if (size % sizeof(uint32_t) ||
+	    !(size > 0)) {
+		return -EINVAL;
+	}
+
+	struct qspi_nor_data *const driver_data = dev->driver_data;
+	const struct qspi_nor_config *params = dev->config->config_info;
+
+	if (driver_data->write_protection) {
+		return -EACCES;
+	}
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((size + addr) > params->size)) {
+		return -EINVAL;
+	}
+
+	qspi_lock(dev);
+
+	int ret = nrfx_qspi_write(src, size, addr);
+
+	qspi_wait_for_completion(dev);
+	return qspi_get_zephyr_ret_code(ret);
+}
+
+static int qspi_nor_erase(struct device *dev, off_t addr, size_t size)
+{
+	struct qspi_nor_data *const driver_data = dev->driver_data;
+	const struct qspi_nor_config *params = dev->config->config_info;
+
+	if (driver_data->write_protection) {
+		return -EACCES;
+	}
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((size + addr) > params->size)) {
```

Comment:
This is also checked in `qspi_erase` in a bit more extensive way, why do it twice?
And `-EINVAL` is returned there, as well as in `qspi_nor_read` and `qspi_nor_write`, which seems more suitable.

Reply:
Right. I reduced this check to one and moved it to the `qspi_nor_erase`.